cmd/thanos/tools_bucket.go

```diff
@@ -1420,3 +1453,90 @@ func registerBucketRetention(app extkingpin.AppClause, objStoreConfig *extflag.P
 		return nil
 	})
 }
+
+func registerBucketUploadBlocks(app extkingpin.AppClause, objStoreConfig *extflag.PathOrContent) {
+	cmd := app.Command("upload-blocks", "Upload blocks push blocks from the provided path to the object storage.")
+
+	tbc := &bucketUploadBlocksConfig{}
+	tbc.registerBucketUploadBlocksFlag(cmd)
+
+	cmd.Setup(func(g *run.Group, logger log.Logger, reg *prometheus.Registry, _ opentracing.Tracer, _ <-chan struct{}, _ bool) error {
+		ctx := context.Background()
+		httpConfContentYaml, err := tbc.prometheus.httpClient.Content()
+		if err != nil {
+			return errors.Wrap(err, "getting http client config")
+		}
+		httpClientConfig, err := httpconfig.NewClientConfigFromYAML(httpConfContentYaml)
+		if err != nil {
+			return errors.Wrap(err, "parsing http config YAML")
+		}
+
+		httpClient, err := httpconfig.NewHTTPClient(*httpClientConfig, "thanos-tool")
+		if err != nil {
+			return errors.Wrap(err, "Improper http client config")
+		}
+
+		m := &promMetadata{
+			promURL: tbc.prometheus.url,
+			client:  promclient.NewWithTracingClient(logger, httpClient, "thanos-tool"),
+		}
+
+		err = runutil.Retry(2*time.Second, ctx.Done(), func() error {
```

Comment:
I think there is usecase to upload blocks without running Prometheus. In this case external labels are specified as a flag.
This can be done in future prs

Reply:
It'd be nice to have a command to get those external labels independently of this process.

I can see an example where you want to migrate blocks from one cluster to another, but you want to retain the labels from the old cluster which might already be shut down.

Reply:
Yeah we should have follow-up where we replace by extflags from commandline I think